src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -277,13 +280,19 @@ async def mypy_typecheck(
     all_used_source_roots = sorted(
         set(itertools.chain(plugin_sources.source_roots, typechecked_sources.source_roots))
     )
+    extra_env = {"PEX_EXTRA_SYS_PATH": ":".join(all_used_source_roots)}
+    # If the constraints are different for the tool than for the requirements, we must tell Pex to
+    # ignore errors. Otherwise, we risk runtime errors about missing dependencies.
+    if code_interpreter_constraints != tool_interpreter_constraints:
+        extra_env["PEX_IGNORE_ERRORS"] = "true"
```

Comment:
Is this because we are adding the code to mypy's python path? Should we be passing it via MYPYPATH/mypy_path instead? https://mypy.readthedocs.io/en/stable/running_mypy.html#unable-to-find-module

If so, this relates to a pattern in the JVM: isolating "the code that is running" from "the code that is being operated on" as much as possible is good, because it avoids the need to "shade". While it sounds like mypy must run using the interpreter that it is checking for, isolating its dependencies from the code it is checking is still good.

Reply:
That could work..definitely worth trying out. Thanks for the suggestion.

Reply:
This works! If we set MYPYPATH to the wheel embedded in the Pex (or saved in the global cache), MyPy will use it. Even if the wheel is not compatible with the interpreter it's run with!

This test loads up a py2-only wheel:

```
MYPYPATH=~/.pex/installed_wheels/3780598d96101a8267782a69e0b66dc1a54f1def/x690-0.2.0-py2-none-any.whl mypy app.py
```

So, now my question is what is the safest and least hacky way for us to get the absolute paths to the requirments.pex's distributions? Keeping in mind complexities like remote execution, and that the chroot could change every run.

1) Extract from the `requirements.pex` itself by looking at its `.deps` folder. I don't know how to safely absolutify this, and maybe we don't need absolute paths as MyPy can use relative. I think we would run `unzip` on it, and then use `Snapshot` to get the file names. Feed the `Digest` into the runner Pex, and set `MYPYPATH` to the relative files?
2) Use the global cache by inspecting the `--named-cache-dirs` option. This sounds much less safe, e.g. issues with remoting.

Reply:
#9206 is the ticket tracking the general case for shading. I had started working on a solution to the above which used a thin runner script to download requirements using the pex library API, and to execute that within a `Process`. I still think that could be a good solution for mypy instead of relying on pex internals. If the `.deps` folder is considered stable enough, I don't see a problem with unzipping it.